File: Meeting Rooms III.py
# https://leetcode.com/problems/meeting-rooms-iii/?envType=daily-question&envId=2024-02-18
# Meeting Rooms III

from heapq import heappush, heappop

class Solution:
    def mostBooked(self, n: int, meetings: List[List[int]]) -> int:

        meetings.sort(key=lambda x: x[0])
    
        free_rooms = [(i, 0) for i in range(n)]
        ongoing_meetings = []
        
        room_usage_count = [0] * n
        
        for start, end in meetings:
            while ongoing_meetings and ongoing_meetings[0][0] <= start:
                _, room = heappop(ongoing_meetings)
                heappush(free_rooms, (room, start))
            
            if free_rooms and free_rooms[0][1] <= start:
                room, _ = heappop(free_rooms)
                room_usage_count[room] += 1
                heappush(ongoing_meetings, (end, room))
            else:
                earliest_end, room = heappop(ongoing_meetings)
                room_usage_count[room] += 1
                new_end = max(earliest_end, start) + (end - start)
                heappush(ongoing_meetings, (new_end, room))

        max_meetings = max(room_usage_count)
        return room_usage_count.index(max_meetings)

File: Meeting Rooms III.js
// https://leetcode.com/problems/meeting-rooms-iii/?envType=daily-question&envId=2025-12-27
// Meeting Rooms III

/**
 * @param {number} n
 * @param {number[][]} meetings
 * @return {number}
 */

var mostBooked = function(n, meetings) {
    meetings.sort((a, b) => a[0] - b[0]);

    const idle = [];

    for (let i = 0; i < n; i++) idle.push(i);
    idle.sort((a, b) => a - b);

    const busy = [];

    const cnt = Array(n).fill(0);

    const pushBusy = (item) => {
        busy.push(item);
        busy.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
    };

    const pushIdle = (room) => {
        idle.push(room);
        idle.sort((a, b) => a - b);
    };

    for (const [s, e] of meetings) {
        while (busy.length && busy[0][0] <= s) {
            const [, room] = busy.shift();
            pushIdle(room);
        }

        if (idle.length) {
            const room = idle.shift();
            cnt[room]++;
            pushBusy([e, room]);
        } else {
            const [endTime, room] = busy.shift();
            cnt[room]++;
            pushBusy([endTime + (e - s), room]);
        }
    }

    let ans = 0;

    for (let i = 1; i < n; i++) {
        if (cnt[i] > cnt[ans]) ans = i;
    }

    return ans;
};